### PR TITLE
Assignment 2

### PR DESCRIPTION
**Pull request by ThomasdenH**
My submission for [Assignment 2](https://github.com/yieldprotocol/mentorship2022/issues/2).

The contract is deployed on Rinkeby [here](https://rinkeby.etherscan.io/address/0xdd2505c199923975881bdb5fde467013c07fdc3b). For some reason, verification didn't work. My first try was with the `--verify` flag using `forge create` which deployed the contract but couldn't verify so my guess is that it's either a bug in Forge or with Etherscan. Manual verification didn't work either, so maybe something with Etherscan?

I can rebase into one commit if that's necessary.

### CODE REVIEW

File: Assignment 2/src/Coll.sol [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import "yield-utils-v2/mocks/ERC20Mock.sol";
+
+/// @notice `Coll`, the ideal token to use as collateral for vaults! At least
```
Comment (devtooligan) [inline Assignment 2/src/Coll.sol:6]: 🤣

File: Assignment 2/src/Vault.sol [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    ///     withdrawn.
+    event Withdrawal(address indexed account, uint256 withdrawn);
+
+    /// @notice Deposits made using `deposit` can be reverted with this error
+    ///     if the transfer failed.
+    /// @dev This error is thrown if `transferFrom` returns `false`. Instead of
+    ///     returning false however, it is possible that the `IERC20` reverts,
+    ///     for example when attempting to transfer more than the allowance.
+    ///     So: not all failed deposits will throw this error.
+    error DepositFailed(address account, uint256 amount);
```
Comment (devtooligan) [inline Assignment 2/src/Vault.sol:35]: I haven't seen these type of failed events before but sounds potentially valuable.  Is there any reason u did not6 index the account here or on the next one?

Comment (devtooligan) [inline Assignment 2/src/Vault.sol]: I'm not used to seeing errors here so I got confused and thought these were events at first.  I would suggest either moving them outside the contract to the top level (on larger ctrcts w a lot of errors, they can even have their own file) or else using section headers for each section within the contract

File: Assignment 2/src/Vault.sol [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        depositOf[msg.sender] += amount;
+        // Transfer the tokens to this contract
+        bool successful = token.transferFrom(msg.sender, address(this), amount);
+
+        // Finally, check if the transfer was successful to either revert with
+        // an error or emit an event.
+        if (successful) {
+            emit Deposit(msg.sender, amount);
+        } else {
+            revert DepositFailed(msg.sender, amount);
```
Comment (devtooligan) [inline Assignment 2/src/Vault.sol:73]: its definitely a graceful way to handle errors.  i usually just let it revert lol

File: Assignment 2/src/Vault.sol [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    ///     withdrawn is more than the balance of the address. It will also
+    ///     fail if the `transfer` function returns `false`.
+    /// @param amount The amount to withdraw.
+    /// @dev Follows the Checks-Effects-Interactions Pattern.
+    ///
+    ///     The `else` branches could be combined by adding a `return`, but to
+    ///     keep the control flow as obvious as possible they aren't.
+    function withdraw(uint256 amount) external {
+        if (depositOf[msg.sender] >= amount) {
+            depositOf[msg.sender] -= amount;
```
Comment (devtooligan) [inline Assignment 2/src/Vault.sol]:
```suggestion
        deposit_ = depositOf[msg.sender] // saves 2 sloads below
        if (deposit_ >= amount) {
            depositOf[msg.sender] = deposit_ - amount;
```

File: Assignment 2/test/Vault.t.sol [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+/// @notice Test state where the contract has minted some tokens.
+abstract contract MintedTokenState is ZeroState {
+    /// @notice The amount of tokens that the contract should own at the start
+    ///     of tests.
+    uint256 constant tokensMinted = 1_000_000_000_000_000_000;
+
+    function setUp() public virtual override {
+        super.setUp();
+        token.mint(address(this), tokensMinted);
```
Comment (devtooligan) [inline Assignment 2/test/Vault.t.sol]: ur using this test contract as the user.  a more clear pattern is to setup a user address in state and refer to it.  everytime u want that user to do something, use vm.prank.   see [calnix example](https://github.com/calnix/Basic-Vault/pull/2/files#diff-63defda6a33f4b1b264b918d2690459a930533f21cd6f513259ea2b7cb632f5bR23-R24)

it avoids confusion, and becomes more useful when u have more than one user

File: Assignment 2/test/Vault.t.sol [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+abstract contract DepositedTokenState is MintedTokenState {
+    function setUp() public override {
+        super.setUp();
+        token.approve(address(vault), tokensMinted);
+        vault.deposit(tokensMinted);
+    }
+}
+
+/// @notice Test some simple properties of the contract in the `ZeroState`.
+contract ZeroStateTest is ZeroState, Test {
```
Comment (devtooligan) [inline Assignment 2/test/Vault.t.sol]: if u make Zerostate inherit from Test then u only need to inherit from ZeroState

**Review by devtooligan — APPROVED** [review]
looks great! just some smoltings to think about.  approved

**Comment by ThomasdenH** [comment]
Thanks for the review! I implemented the suggestions.

**Pull request merged**